contracts/System.sol

```diff
@@ -80,4 +121,175 @@ contract System {
   /// @param lowerBound requested lower bound of the param.
   /// @param upperBound requested upper bound of the param
   error OutOfBounds(string name, uint256 given, uint256 lowerBound, uint256 upperBound);
+
+  struct Addresses {
+    address validatorSet;
+    address slash;
+    address systemReward;
+    address lightClient;
+    address relayerHub;
+    address candidateHub;
+    address govHub;
+    address pledgeAgent;
+    address burn;
+    address foundation;
+  }
+
+/* @dev ExtStorage: an extended storage struct used to store base contract state variables in a 'floating' 
+   i.e. non-sequential storage slot. this is done to avoid collisions with derived contracts' state variables.       
+   Uses ERC-7201 namespace to reduce chances of collision, see https://eips.ethereum.org/EIPS/eip-7201
+  */
+  struct ExtStorage {
+    /// @custom:storage-location erc7201:core.system.extended.storage
+    Addresses addrs;
+    uint256 guardStatus;
+    // @dev additional extended-storage fields goes here
+  }
+
+  // ERC-7201 namespace: keccak256(abi.encode(uint256(keccak256("core.system.extended.storage")) - 1)) & ~bytes32(uint256(0xff));
+  bytes32 private constant _EXT_STORAGE_LOCATION = 0x67a7102a872f79141e3e6e107bddf1a70e311104f2c8c3897ed1c7c60974e600;
+
+  constructor() {
+    s_testModeDeployer = _isLocalTestNode() ? msg.sender : address(0);
+  }
+
+/* @dev _ext(): obtain a storage reference to the non-sequential extended storage struct
+   see https://solidity.readthedocs.io/en/v0.8.4/internals/layout_in_storage.html#layout-of-state-variables-in-storage
+  */
+  function _ext() private pure returns (ExtStorage storage $) {
```

Comment:
Gokberk: please review the way I've implemented non-sequential struct storage and ERC-7201 namespace

Reply:
I see now why you wanted Halborn's opinion here. It's a required external check before we put these changes/the new standards they herald into production.

Once we've merged this and are confident with all the appropriate tests, I agree that we should present this to Halborn for an audit. Not the entire set of changes - but specifically the core architectural lynchpins such as non-sequential struct storage and ERC-7201.

We can prepare the 'best practices' and/or 'code standards' guides in preparation for the external review as well

Reply:
Yep. The only thing I want Halburn to review is my usage of extended 'floating' storage to solve collisions.

Reply:
> assign a fixed storage slot to it. 

basically yes. the key factor here is that this new storage is non-sequential since it is not implemented as state variable but rather as a function returning a storage pointer.

> Theoretically, there is still very slight chance of storage collision

which is also the case for map entries whose location is set by solidity to be the kaccak of their keys and of dynamic array whose values' location is also some hashing product. ERC-7201 is said to reduce even such a theoretical risk by applying an additional 'scattering' logic over keccak